File: src/main/java/me/andre111/d20server/command/SayCommand.java
package me.andre111.d20server.command;

import me.andre111.d20common.model.chat.ChatEntry;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.scripting.Context;
import me.andre111.d20common.scripting.ScriptException;
import me.andre111.d20common.scripting.variable.Variable;
import me.andre111.d20common.scripting.variable.parser.VariableParser;
import me.andre111.d20server.service.ChatService;

public class SayCommand extends Command {
	public SayCommand(String name, String[] aliases) {
		super(name, aliases);
	}

	@Override
	public void execute(Profile profile, String arguments) {
		try {
			Variable variable = VariableParser.parseVariable("selected.property.name");
			Object name = variable.get(new Context(profile, profile.getMap(), null));
			
			StringBuilder sb = new StringBuilder();

			sb.append("<p class=\"chat-sender\">");
			sb.append(ChatService.escape(""+name));
			sb.append(" (");
			sb.append(ChatService.escape(profile.getName()));
			sb.append("): ");
			sb.append("</p>");
			sb.append("<p class=\"chat-message\">");
			sb.append(ChatService.escape(arguments));
			sb.append("</p>");
			
			ChatService.append(true, new ChatEntry(sb.toString(), profile.id()));
		} catch (ScriptException e) {
			ChatService.appendError(profile, "Could not send message:", e.getMessage());
		}
	}
}

File: core/server/command/say-command.js
// @ts-check
import { Command } from './command.js';
import { ChatService } from '../service/chat-service.js';

import { ChatEntry } from '../../common/message/chat/chat-entry.js';
import { Scripting } from '../../common/scripting/scripting.js';

const SCRIPT = new Scripting(false);
const EXPR = SCRIPT.parseExpression('sActor.name');

export class SayCommand extends Command {
    constructor(name, aliases) {
        super(name, aliases, false);
    }

    execute(profile, args) {
        const name = SCRIPT.evalExpression(EXPR, profile, null);
        SCRIPT.throwIfErrored();

        // parse message
        const parsed = ChatService.parseInlineExpressions(args);

        var text = '<div class="chat-sender">' + ChatService.escape('' + name.value) + ' (' + ChatService.escape(profile.getUsername()) + '): </div>';
        text = text + '<div class="chat-message">' + parsed.string + '</div>';

        ChatService.append(true, new ChatEntry(text, profile.getID(), true, null, parsed.diceRolls, parsed.triggeredContent));
    }
}
